### PR TITLE
Added AsReadOnly() extension to ISet<T>

### LINKED ISSUE #29387: Add an `ISet<T>.AsReadOnly()` extension method
The [`List<>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1) class have a [`AsReadOnly`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.asreadonly) method to turn the collection into a [`ReadOnlyCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1), but [`HashSet<>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1) does not have any such method.

## Rationale and usage
Allows Entity Framework Core to expose a [collection navigation property](https://docs.microsoft.com/en-us/ef/core/modeling/relationships) as a [`IReadOnlyCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1) when backing it with a private [backing field](https://docs.microsoft.com/en-us/ef/core/modeling/backing-field) declared as a `HashSet`.
```cs
[Table("Blogs")]
public class Blog
{
    private HashSet<Post> _posts;
 
    public int BlogId { get; set; }
 
    public IReadOnlyCollection<Post> Posts => _posts?.AsReadOnly();
 
    public void AddPost(Post post)
    {
        // Do some business logic here...
        _posts.Add(post);
    }
}
```

## Proposed API
```diff
namespace System.Collections.Generic;

public class CollectionExtensions
{
      public ReadOnlyCollection<T> AsReadOnly(this IList<T> list);
      public ReadOnlyDictionary<T> AsReadOnly(this IDictionary<T> dictionary);
+     public ReadOnlySet<T> AsReadOnly(this ISet<T> set);
}

### PR DESCRIPTION
**Pull request by Mrxx99**
this adds the asReadOnly extension method to ISet<T>
unfortunately CollectionExtensions does not find ReadOnlySet, I guess it either has to be moved to System.Private.CoreLib and/or type forwarders have to be added. But I don't know how to do that.

fixes #29387

### CODE REVIEW

**Comment by stephentoub** [comment]
> I guess it either has to be moved to System.Private.CoreLib

Yes, the implementation of `ReadOnlySet<T>` would need to be moved down to corelib. From a reference assembly perspective, it's in `System.Collections.dll` along with `CollectionExtensions`, so it's only an implementation issue. You can just move the implementation down.

**Comment by Mrxx99** [comment]
>  You can just move the implementation down.

Unfortunately I still get the following error:
```
C:\Code\Other\dotnet\runtime\src\libraries\System.Private.CoreLib\src\System\Collections\Generic\CollectionExtensions.cs(76
,23): error CS0246: The type or namespace name 'ReadOnlySet<>' could not be found (are you missing a using directive or an
assembly reference?) [C:\Code\Other\dotnet\runtime\src\coreclr\nativeaot\System.Private.CoreLib\src\System.Private.CoreLib.
csproj]
```

**Comment by stephentoub** [comment]
> Unfortunately I still get the following error:

This is editing the wrong System.Private.CoreLib project file. You instead want to add it to https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems.

CoreLib for coreclr, mono, and native aot is slightly different for each. Most of the source is the same, and that shared source is listed in the above projitems file. The file you edited is the one that contains just the additional source specific to coreclr.

File: src/libraries/System.Collections/ref/System.Collections.cs [hunk 1]
```diff
@@ -580,6 +580,8 @@ public partial struct Enumerator : System.Collections.Generic.IEnumerator<T>, Sy
         }
     }
 }
+#endif // !BUILDING_CORELIB_REFERENCE
+
 namespace System.Collections.ObjectModel
 {
     public partial class ReadOnlySet<T> : System.Collections.Generic.ICollection<T>, System.Collections.Generic.IEnumerable<T>, System.Collections.Generic.IReadOnlyCollection<T>, System.Collections.Generic.IReadOnlySet<T>, System.Collections.Generic.ISet<T>, System.Collections.ICollection, System.Collections.IEnumerable
```
Comment (eiriktsarpalis) [inline src/libraries/System.Collections/ref/System.Collections.cs:587]: Seeing as the other `System.Collections.ObjectModel` collections that are already in System.Private.CoreLib are not listed in this file, does `ReadOnlySet<T>` still need to be here?

**Comment by eiriktsarpalis** [comment]
@stephentoub @artl93 can we make an exception and slot this API into .NET 9? It's low risk and complements the newly introduced `ReadOnlySet<T>`.

File: src/libraries/System.Collections/ref/System.Collections.cs (hunk 1 above)
Comment (Mrxx99) [reply src/libraries/System.Collections/ref/System.Collections.cs:587]: When I remove it here, I get this error when running `build.cmd clr+libs -rc Release`:
```
C:\Code\Other\dotnet\runtime\src\libraries\System.Collections\ref\System.Collections.cs(600,54): error CS0234: The type
 or namespace name 'ReadOnlySet<>' does not exist in the namespace 'System.Collections.ObjectModel' (are you missing an
 assembly reference?) [C:\Code\Other\dotnet\runtime\src\libraries\System.Private.CoreLib\ref\System.Private.CoreLib.csp
roj]
```

Comment (eiriktsarpalis) [reply src/libraries/System.Collections/ref/System.Collections.cs:587]: It should be added to the System.Runtime reference project:

https://github.com/dotnet/runtime/blob/07f984305205a760edd2ace52670eefdaf3072a2/src/libraries/System.Runtime/ref/System.Runtime.cs#L8367-L8399

Here are instructions on how to regenerate the reference file:

https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md

Comment (Mrxx99) [reply src/libraries/System.Collections/ref/System.Collections.cs]: I had no luck with these instructions as builds failed as soon as I removed it from the System.Collections ref project. So I added it manually to System.Runtime ref project.

File: src/libraries/System.Collections/tests/Generic/CollectionExtensionsTests.cs [hunk 2]
```diff
@@ -134,6 +143,13 @@ public void AsReadOnly_NullIList_ThrowsArgumentNullException()
             Assert.Throws<ArgumentNullException>("list", () => list.AsReadOnly());
         }
 
+        [Fact]
+        public void AsReadOnly_NullISet_ThrowsArgumentNullException()
+        {
+            ISet<string> set = null;
+            Assert.Throws<ArgumentNullException>("set", () => set.AsReadOnly());
```
Comment (stephentoub) [inline src/libraries/System.Collections/tests/Generic/CollectionExtensionsTests.cs:150]: Nit:
```suggestion
            AssertExtensions.Throws<ArgumentNullException>("set", () => set.AsReadOnly());
```
Some size-focused builds may strip out the parameter name.

**Review by stephentoub — APPROVED** [review]
Thanks

**Comment by GerardSmit** [comment]
Sorry if I'm misunderstanding this, but if the type is moved from `System.Collections` to `System.Private.CoreLib` and this is landing in .NET 10 instead of .NET 9, shouldn't there be a TypeForward?

**Comment by Mrxx99** [comment]
@stephentoub Is there something missing that is holding this back?

**Comment by eiriktsarpalis** [comment]
This looks ready to merge. I've updated the PR branch to give this another run with more recent bits.

**Comment by stephentoub** [comment]
> This looks ready to merge. I've updated the PR branch to give this another run with more recent bits.

I believe it will need to updated with type forwards, as suggested in https://github.com/dotnet/runtime/pull/106037#issuecomment-2333776296.

**Comment by eiriktsarpalis** [comment]
@Mrxx99 when possible, could you take a look at the outstanding issue?

**Comment by Mrxx99** [comment]
> @Mrxx99 when possible, could you take a look at the outstanding issue?

I can take a look but I am not familiar with the type forwarders so that could be much work and still would need to be reviewed by one knowledgeable person anyway. Maybe it would be easier if some expert would do it directly.

**Review by stephentoub — APPROVED** [review]

**Comment by stephentoub** [comment]
Thanks

**Comment by Mrxx99** [comment]
@stephentoub Thanks for finishing the PR

**Pull request merged**